sound/soc/sof/intel/hda-dsp.c

```diff
@@ -442,7 +442,13 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 	/* fallback to the previous register value */
 	value = value ? 0 : SOF_HDA_VS_D0I3C_I3;
 
-	return hda_dsp_update_d0i3c_register(sdev, value);
+	/*
+	 * This can fail but return the IPC error to signal that
+	 * the state change failed.
+	 */
+	hda_dsp_update_d0i3c_register(sdev, value);
```

Comment:
@lyakh this is already done in hda_dsp_update_d0i3c_register()? Do you want to print another one?

Reply:
are those fixes needed or just nice-to-have? Did you actually see this fail?

Reply:
@plbossart yes, I did see it fail. On the CML chromebook the D0i3 entry IPC resulted in an error but the new state was changed nevertheless without this change.

Reply:
and what happens then even with this change, you return an IPC error and it's the end?

Reply:
@plbossart for example lets assume we're requesting to enter D0I3 and the IPC fails. Without this change, returning the return value of hda_dsp_update_d0i3c_register() (which could be 0) would indicate that the state change was successful. This would result in hda_dsp_set_power_state() updating the current DSP state to the requested change when in fact it never went through. But if we return an error, the DSP state remains unchanged and the error will be propogated to the caller of the set_power_state() op.

Reply:
and what happens then when you propagate the error further? do you retry a D0i3 transition later? Never?

Reply:
No we do not retry the transition. In all probability based on previous results, if the IPC fails once, it will likely fail again. So even if we retry it, I dont think it will help.
But the D0i3_work could get scheduled again after a recent IPC and that would result in retrying.

Reply:
you should definitely add this reply as a comment, it's very useful information

Reply:
@plbossart updated now. Is this better?